### PR TITLE
Fixes a bug in SESERegion Builder

### PR DESCRIPTION
**Pull request by bgogul**
The computation of true and false regions of a conditional region should only consider nodes up to the post idom of the current startBB. Currently, it considers all the way up to the endBB, which causes duplication of the blocks in the SESE regions.

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Review by rxwei — COMMENTED** [review]
Any test case where the original case would trigger a crash or failure?

**Comment by bgogul** [comment]
@rxwei, nope there were no test cases for SESE construction. I was planning to add new test cases in  a subsequent PR.

**Comment by mhong** [comment]
would it be useful to add one as part of this PR, as a holistic piece of change?

It'd also be nice to grep (use CHECK) on "XLA CFG Canonicalize" to confirm we are creating the right CFG.

**Comment by bgogul** [comment]
@mhong @rxwei Added a simple test with multiple ifs. PTAL.

File: test/TensorFlow/integration.swift [hunk 1]
```diff
@@ -228,6 +228,65 @@ public func test_bool_param2(cond: Bool, // expected-warning {{'cond' implicitly
 // CHECK-NEXT: [[PROGRAM:%.*]] = apply [[STARTFN:%.*]](
 // CHECK: cond_br [[BOOLVAL]],
 
+public func test_multiple_ifs(status: Bool // expected-warning {{'status' implicitly copied to the accelerator}}
+) {
+  var a = Tensor<Int32>(0);
```
Comment (rxwei) [inline test/TensorFlow/integration.swift]: Swift doesn't need semicolons or parentheses around `if` conditions.

File: test/TensorFlow/integration.swift [hunk 2]
```diff
@@ -228,6 +228,65 @@ public func test_bool_param2(cond: Bool, // expected-warning {{'cond' implicitly
 // CHECK-NEXT: [[PROGRAM:%.*]] = apply [[STARTFN:%.*]](
 // CHECK: cond_br [[BOOLVAL]],
 
+public func test_multiple_ifs(status: Bool // expected-warning {{'status' implicitly copied to the accelerator}}
```
Comment (rxwei) [inline test/TensorFlow/integration.swift]: To put the expected-warning on top and put the `) {` back on this line, you can write:
```swift
// expected-warning @+1 {{'status' implicitly copied to the accelerator}}
public func test_multiple_ifs(status: Bool)
``

File: test/TensorFlow/integration.swift [hunk 3]
```diff
@@ -228,6 +228,65 @@ public func test_bool_param2(cond: Bool, // expected-warning {{'cond' implicitly
...
+  let b = a;
+  if (status) { // expected-note {{value used here}}
+    a += b;
+  }
+  a += b;
+  if (status) {
+    a += b;
+  }
+  a -= b;
+  print (a)
```
Comment (rxwei) [inline test/TensorFlow/integration.swift]: No need for space after a function name.

Comment (mhong) [inline test/TensorFlow/integration.swift]: use _hostOp(a) so that the generated SIL code is cleaner

File: test/TensorFlow/integration.swift [hunk 4]
```diff
@@ -228,6 +228,65 @@ public func test_bool_param2(cond: Bool, // expected-warning {{'cond' implicitly
...
 // CHECK: cond_br [[BOOLVAL]],
 
+public func test_multiple_ifs(status: Bool // expected-warning {{'status' implicitly copied to the accelerator}}
+) {
+  var a = Tensor<Int32>(0);
+  let b = a;
+  if (status) { // expected-note {{value used here}}
+    a += b;
+  }
+  a += b;
```
Comment (mhong) [inline test/TensorFlow/integration.swift]: do we need this statement in the reproducer?

File: test/TensorFlow/integration.swift [hunk 5]
```diff
@@ -228,6 +228,65 @@ public func test_bool_param2(cond: Bool, // expected-warning {{'cond' implicitly
...
+  if (status) {
+    a += b;
+  }
+  a -= b;
+  print (a)
+}
+
+// CHECK-LABEL: --- TFPartition Accelerator Result: {{.*}}test_multiple_ifs{{.*}}
```
Comment (mhong) [inline test/TensorFlow/integration.swift]: Does CHECKing "TFPartition Accelerator Result" in this test add value? It seems CHECKing on "XLA CFG Canonicalize" would be sufficient. If we must, maybe also check on the presence of "cond_br".

Minimizing the CHECK'ed content would help reduce test maintenance efforts.

e.g. when we later add a new attr to builtin "__tfop_Const", or migrate it to the new graph op inst, or rename "tf_tensor_to_i1" to something else, etc.

**Review by mhong — APPROVED** [review]
Nice fix! Should we also add some general sanity code to make sure each BB is lowered exactly once (and appears in exactly one SESE region)?

Please file a TODO or bug to track that.

File: test/TensorFlow/integration.swift (hunk 5 above)
Comment (bgogul) [reply test/TensorFlow/integration.swift]: Good point. I simplified the test. I will also move this over later to a separate file that checks canonicalization and SESE region construction.

File: test/TensorFlow/integration.swift (hunk 3 above)
Comment (bgogul) [reply test/TensorFlow/integration.swift]: Thanks!

File: test/TensorFlow/integration.swift (hunk 2 above)
Comment (bgogul) [reply test/TensorFlow/integration.swift]: Ah, that is nice. Thanks for the tip. :)

File: test/TensorFlow/integration.swift (hunk 1 above)
Comment (bgogul) [reply test/TensorFlow/integration.swift]: Done.

File: test/TensorFlow/integration.swift (hunk 4 above)
Comment (bgogul) [reply test/TensorFlow/integration.swift]: Yes, this is needed. Otherwise, I cannot reproduce the bug.

**Review by rxwei — APPROVED** [review]
LGTM

**Pull request merged**